### PR TITLE
WIP: indent on typing dot

### LINKED ISSUE #439: Reindent chain calls on typing dot.
Consider this code:

```rust
    /// Finds a child module with the specified name.
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)|
    }
```

| is the caret position
if I press <kbd>Enter</kbd>, I get

```rust
    /// Finds a child module with the specified name.
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
        |
    }
```

which is reasonable.

however, if I then press <kbd>.</kbd>, I get

```rust
    /// Finds a child module with the specified name.
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
        .|
    }
```

while I really would love to get this:

```rust
   /// Finds a child module with the specified name.
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .|
    }
```

To achive this effect, we should add an `on_dot_typed` function which receives the input text (with the dot) and the dot position, and which returns a text edit which should add a level of indent. 

The LSP-protocol level entry here is this function:

https://github.com/rust-analyzer/rust-analyzer/blob/3e42a158787955ff9f2e81be43479dbe8f2b1bb6/crates/ra_lsp_server/src/main_loop/handlers.rs#L91-L97

It currently uses a similar logic to automatically add `;` once you've typed `let var = ` before an expression.

this function calls into `ra_analysis`, which is a fronted for all clever stuff. So, `on_dot_typed` should be added over here:

https://github.com/rust-analyzer/rust-analyzer/blob/3e42a158787955ff9f2e81be43479dbe8f2b1bb6/crates/ra_analysis/src/lib.rs#L352-L356

and call into `on_eq_type` function in `ra_editor` (a crate which is restricted to working with a single file) here:

https://github.com/rust-analyzer/rust-analyzer/blob/3e42a158787955ff9f2e81be43479dbe8f2b1bb6/crates/ra_editor/src/typing.rs#L108

The tests should go into `typing.rs` as well.

As for how to actually determine the apropriate indent, I am not sure! But a simple heuristic like "check that a line is all whitespace and a dot, figure the indent on the previous line, indent by four spaces" should work pretty well I guess, and tweaking the heuristic *after* all the plumbing is done and there are unit-tests should be a breeze!

### PR DESCRIPTION
**Pull request by simonvandel**
Fixes #439.

The unit test passes, but I can't seem to make VS code perform the action. The existing action on "=" doesn't work either on my end either though.

I didn't add any smart way of detecting the current indent level. Any ideas how I would do that?

### CODE REVIEW

**Comment by max-frai** [comment]
>The existing action on "=" doesn't work either on my end either though.
Just in case. Do you have installed VIM extension? If yes, then enhanced typing is disabled at all.

**Comment by simonvandel** [comment]
> > The existing action on "=" doesn't work either on my end either though.
> 
> Just in case. Do you have installed VIM extension? If yes, then enhanced typing is disabled at all.

I'm not sure what was going on. In any case, the current state of the branch has the existing "=" action working, but the "." action does not. I'm not seeing any "onTypeFormatting" events being sent to the server, so my guess is that something else is eating the "." typing.

**Comment by matklad** [comment]
I think that problem is that we need to tell vscode upfront about .: 
https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_lsp_server/src/caps.rs#L38

**Comment by simonvandel** [comment]
> I think that problem is that we need to tell vscode upfront about .:
> https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_lsp_server/src/caps.rs#L38

Yep, that seems to work. Thanks @matklad

File: crates/ra_editor/src/typing.rs [hunk 1]
```diff
@@ -614,6 +637,31 @@ fn foo() {
...
+",
+            r"
+    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
+        self.child_impl(db, name)
+            .
+    }
+",
+        );
+    }
+
```
Comment (matklad) [inline crates/ra_editor/src/typing.rs]: A several more tests I think we should have is:

* don't reindernt if the dot is already indented. That is, edit should be None in this situation

```
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .<|>
    }
```

* indent if the previous line is already indented

```
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .first()
        .<|>
    }
```

->


```
    pub fn child(&self, db: &impl HirDatabase, name: &Name) -> Cancelable<Option<Module>> {
        self.child_impl(db, name)
            .first()
            .<|>
    }
```

3) the combination of 1 & 2: no indent if dot is already indented and there's an indent on the previous line.

File: crates/ra_editor/src/typing.rs [hunk 2]
```diff
@@ -136,6 +136,27 @@ pub fn on_eq_typed(file: &SourceFileNode, offset: TextUnit) -> Option<LocalEdit>
     })
 }
 
+pub fn on_dot_typed(file: &SourceFileNode, offset: TextUnit) -> Option<LocalEdit> {
+    let before_dot_offset = offset - TextUnit::of_char('.');
+
+    let _whitespace = find_leaf_at_offset(file.syntax(), before_dot_offset)
+        .left_biased()
+        .and_then(ast::Whitespace::cast)?;
+
```
Comment (matklad) [inline crates/ra_editor/src/typing.rs]: Here we should check that `.` is a part of chain call expression, and not just some random token, like this:

https://github.com/intellij-rust/intellij-rust/blob/ecc8def8829f86a0a6d20e3b0802ec463cf9174c/src/main/kotlin/org/rust/ide/typing/RsDotTypedHandler.kt#L36


I can't actually came up with a test where indenting dot would be *wrong* (and, indeed, IntelliJ has only positive tests), but it's safe to check this anyway.

File: crates/ra_editor/src/typing.rs [hunk 3]
```diff
@@ -136,6 +136,58 @@ pub fn on_eq_typed(file: &SourceFileNode, offset: TextUnit) -> Option<LocalEdit>
     })
 }
 
+pub fn on_dot_typed(file: &SourceFileNode, offset: TextUnit) -> Option<LocalEdit> {
+    let before_dot_offset = offset - TextUnit::of_char('.');
+
+    let _whitespace = find_leaf_at_offset(file.syntax(), before_dot_offset).left_biased()?;
```
Comment (matklad) [inline crates/ra_editor/src/typing.rs]: `_` as the first symbol of var name means that the var is intentionally unused. We use this var, so we don't need `_` here and elsehwere.

File: crates/ra_editor/src/typing.rs [hunk 4]
```diff
@@ -136,6 +136,58 @@ pub fn on_eq_typed(file: &SourceFileNode, offset: TextUnit) -> Option<LocalEdit>
...
+        .skip(1)
+        .next()?;
+
+    ast::MethodCallExprNode::cast(_method_call)?;
+
+    // find how much the _method call is indented
+    let method_chain_indent = _method_call
+        .ancestors()
+        .skip(1)
+        .next()?
```
Comment (matklad) [inline crates/ra_editor/src/typing.rs]: `.acenstors.skip(1).next()` can be written more succinctly as `.parent()` I think?

**Pull request merged**